ci/eval/release-checks: init

winterqt opened the pull request:
This PR reimplements `pkgs/top-level/nixpkgs-basic-release-checks.nix` using the chunked evaluator function, in order for it to be efficient enough to run in CI.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin (with #406307)
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

winterqt committed:
ci/eval/release-checks: init

This change reimplements `pkgs/top-level/nixpkgs-basic-release-checks.nix`
using the chunked evaluator function, in order for it to be efficient
enough to run in CI.

winterqt committed:
ci/eval: don't evaluate packages marked as broken

We really can't expect packages that are marked as broken to evaluate,
and *especially* not on unsupported platforms.

For context, we were attempting to eval them *past* the broken throw
previously, which caused fun side effects like [0], but given that the
next change will cause us to eval packages regardless of `meta.platforms`
as a sanity check, this is required anyways (e.g. `airtame` needs `libgcc`,
which is `null` on Darwin).

[0]: https://github.com/NixOS/nixpkgs/issues/355847#issuecomment-2878873137

emilazy commented on ci/eval/default.nix:
```diff
@@ -36,26 +36,48 @@ let
 
   attrpathsSuperset =
+    {
+      # Evaluates attrpaths from all supported systems, then takes the unique union of the
+      # results. This isn't the default because it doesn't match Hydra's behavior, but this
+      # can be useful for testing sets whose existence is gated on some platform. For
+      # example, strict mode would have caught https://github.com/NixOS/nixpkgs/pull/406303.
+      evalSystems ? [ "x86_64-linux" ],
```
Note that Hydra gathers attrpaths on `x86_64-linux` only, which results in various weird behaviour. I don’t think we need to match that in CI *unless* it can result in a channel‐blocking eval issue that otherwise wouldn’t be caught, though. (I don’t know whether one is possible.)

Some indication that conditioning attrpath existence on platform ruins your chances for Hydra to build your package would be nice, but is not essential.

emilazy replied:
As in: if you gather attrpaths on the same platform you eval them on, then you evaluate strictly fewer things than Hydra, because you will omit attrpaths that are present on `x86_64-linux` for other platforms. I don’t know whether the eval problems that can arise at that point can have any effect other than breaking individual packages that aren’t meant to work on that platform anyway, though.

winterqt replied:
> Some indication that conditioning attrpath existence on platform ruins your chances for Hydra to build your package would be nice, but is not essential.

A comment, or detecting it in the job somehow?

emilazy replied:
I mean that matching Hydra eval semantics with `x86_64-linux` attrpath gathering “correctly” shows the package as absent if it won’t be built on Hydra. (But this is not so important that we should sacrifice other things, like reasonable perf or catching actual channel‐breaking eval issues, for it.)

winterqt replied:
Not sure I follow -- I don't think we'll catch 100% of channel-breaking issues unless we also follow Hydra's attrpath gathering for x86_64-linux and then check it across all systems, right? I guess that's where the perf argument comes in.

winterqt replied:
First one is in the above outdated diff: https://github.com/NixOS/nixpkgs/pull/406303. (The attrpath in question only existed when evaling on Darwin, so Hydra never caught it.)

Second, this isn't really a breakage, but a good idea of how this can cause issues:
```diff
--- result/release-checks-eval/outpaths.json    1969-12-31 19:00:01.000000000 -0500
+++ result/hydra-emulation-eval/outpaths.json    1969-12-31 19:00:01.000000000 -0500
@@ -16507,9 +16507,6 @@
   "elf-header-real.aarch64-darwin": {
     "out": "/nix/store/yy930m0s4zizybbw1q4jqhsd3ajrcc73-elf-header-2.40"
   },
-  "elf-header.aarch64-darwin": {
-    "out": "/nix/store/yy930m0s4zizybbw1q4jqhsd3ajrcc73-elf-header-2.40"
-  },
   "elf-info.aarch64-darwin": {
     "out": "/nix/store/sc8gph0b1x7yjq5vfd3cszhysg3cvqma-elf-info-0.3.0"
   },
@@ -60053,12 +60050,6 @@
   "libinstpatch.aarch64-darwin": {
     "out": "/nix/store/dxv9p299b6v5vsl2k56l0fz2m9zqf26y-libinstpatch-1.1.6"
   },
-  "libintl.aarch64-darwin": {
-    "doc": "/nix/store/bimazwwhk14v3ckdngcsg7wippgniz6v-gettext-0.22.5-doc",
-    "info": "/nix/store/h1wk3blp70f03kmd1vx247ng2f1ngkll-gettext-0.22.5-info",
-    "man": "/nix/store/h0y6w68qd5ma7310afcmi9cdchx04v9f-gettext-0.22.5-man",
-    "out": "/nix/store/f5787h407j2jajnc64jq6zv0jrn3qz5j-gettext-0.22.5"
-  },
   "libiodbc.aarch64-darwin": {
     "out": "/nix/store/whbjvnmvsz2i9wzx6gql9db4g0qma656-libiodbc-3.52.16"
   },
@@ -194681,9 +194672,6 @@
   "tests.pkg-config.defaultPkgConfigPackages.gnome-keyring-1.aarch64-darwin": {
     "out": "/nix/store/0m19wz0862ydvm4x0aid880zhdf65ci4-check-pkg-config-libgnome-keyring"
   },
-  "tests.pkg-config.defaultPkgConfigPackages.hidapi.aarch64-darwin": {
-    "out": "/nix/store/0q8cgg2qpm57660yrqk196x5i67lwx8j-check-pkg-config-hidapi"
-  },
   "tests.pkg-config.defaultPkgConfigPackages.icu-i18n.aarch64-darwin": {
     "out": "/nix/store/va0vymwzp20936n1n6x6c5wwlv9j37pa-check-pkg-config-icu4c"
   },
```

This is the result of the full coverage eval for `aarch64-darwin`. Notice how some attrs just don't exist when Hydra evals it?

winterqt replied:
Indeed, and that's **what we're doing with the changes in this PR**. However, it's still valuable to be able to test what Hydra would see, so I provide the option when running locally.

winterqt commented:
Did some tests:
- union: https://github.com/winterqt/nixpkgs/actions/runs/15006316097
- no union: https://github.com/winterqt/nixpkgs/actions/runs/15006621903

Time difference compared to the OG action is negligible either way (and I suspect a good chunk of that difference in the no union case is 10s of seconds for copying Nixpkgs to the store twice).

winterqt commented:
Just realized I unintentionally implemented a check incorrectly, let me fix it. It'll give a much larger perf hit though. :(

winterqt commented:
Alright so average per system is ~6m40s with the actually correct check: https://github.com/winterqt/nixpkgs/actions/runs/15007089150

winterqt commented on .github/workflows/eval.yml:
```diff
@@ -19,6 +19,42 @@ jobs:
   get-merge-commit:
     uses: ./.github/workflows/get-merge-commit.yml
 
+  release-checks:
```
Split to a separate job for now.

winterqt replied:
Non-release-checks eval is gone now.

emilazy commented:
IMO the ~doubled eval time here is worth it for peace of mind that we’re catching issues upfront, but it does make it important to replace the existing eval job entirely so that we’re not trashing our concurrency by doubling the number of jobs and having half of them take twice as long.

winterqt committed:
OWNERS: add winterqt to `.github/workflows` and `ci/`

I feel like I've done enough work on this that I can effectively review
any changes going forward. :)

winterqt commented:
Trying to figure out how ^ happened.

winterqt commented:
Might be worth adding more details to the commit messages?

winterqt commented:
Added some more documentation of the options added/set and why we want them. Let me know about the commit message(s), otherwise, I think this is good to go.

philiptaron commented on ci/OWNERS:
```diff
@@ -19,7 +19,7 @@
 /.github/workflows/codeowners-v2.yml    @infinisil @wolfgangwalther
 /.github/workflows/nixpkgs-vet.yml      @infinisil @philiptaron @wolfgangwalther
-/ci                                     @infinisil @philiptaron @NixOS/Security @wolfgangwalther
+/ci                                     @infinisil @philiptaron @NixOS/Security @winterqt @wolfgangwalther
```
Awesome. 💕

philiptaron commented on ci/eval/default.nix:
```diff
@@ -36,9 +36,14 @@ let
 
   attrpathsSuperset =
+    {
+      # To match Hydra
```
Is there a reference or link we could add here that provides the full citation?

winterqt replied:
There isn't a citation, but it's just because https://hydra.nixos.org is x86_64-linux. I guess I could clarify that I'm talking about h.n.o specifically?

philiptaron replied:
Yeah, maybe a h.n.o configuration if it's checked into a public repo somewhere?

winterqt replied:
Just for context: this default is just so that things like `eval.full` match Hydra, it's overridden during the actual release checks/eval CI.

winterqt replied:
How's that for documentation? (See `ci/eval/README.md` and `ci/eval/default.nix`.)

vcunat replied:
Hydra.nixos.org setup is here and around: https://github.com/NixOS/infra/blob/main/build/hydra.nix

philiptaron reviewed (commented):
I didn't review the checks file (actual meat of the PR) in depth, but all the surrounding wiring looks good to me.

winterqt committed:
ci/eval/compare: don't die if there aren't enough stats to compare

Ran into this when comparing two `quickTest` evals.

winterqt commented on ci/eval/release-checks.nix:
```diff
@@ -0,0 +1,127 @@
+      touch "$out"
+    '';
+
+    attrpathsPurity = runCommand "attrpaths-purity-check" { } ''
```
Probably not very often, but given this is still faster than OfBorg by an order of magnitude (start to finish is around 6-7 minutes), I don’t believe that this is too concerning.

winterqt replied:
IMO (and I’d imagine others would agree?), if CI takes just a bit longer to avoid blocking the channels or other shenanigans, it’s worth it. But 🤷‍♀️. Open to discussing it more.

Mic92 replied:
To it feels that doubling the CI time worse than having to occasionally fix this issue. I personally have not encountered sometime adding this issue to any pull request I reviewed. In both cases were not actually producing any harm as the channel is not advancing. But it feels like waiting longer for github CI to finish waste more human time than fixing a rare issue. Can we maybe move this check somehow to a different place i.e. something that doesn't run on every pull request. I have the feeling it would be enough to run this just once an hour.

Mic92 replied:
cc @vcunat and @mweinelt who deal more with channel bumps than me.

winterqt replied:
I’d be open to that, but also… it’s just three minutes more.

Obviously, that will increase over time, but I’m not sure it’s worth fussing too much over now — we have a much faster evaluation strategy than e.g. nixpkgs-review, and people run that/wait for its results on a bunch of PRs before reviewing and/or merging.

So I think we should consider just doing it, and disabling it if it’s that bad, but I really don’t think it’s worth much in the grand scheme of things right now.

Mic92 replied:
nixpkgs-review is actually waiting for this github action now by default unless you are not providing a GITHUB_TOKEN. i.e. for the reason that not everyone has as much spare ram for evaluation than what we have in github actions. I would like to do a counter offer. A github action that runs as a cron task and would git-bisect to the pull request causing this issue.

Mic92 replied:
It's not just for reviwers but also for contributors. It quite often happen that people open a pull request, than run nixpkgs-review only to find out that they need to fix more packages, which result in new changes to be pushed to a pull request. I don't want to double the time they need to wait in this case to get the new evaluation list.

winterqt replied:
I’m not sure I follow, wouldn’t all of the issues be fixed by that second push? That is:
1. user opens PR
2. eval runs
3. user runs nixpkgs-review
4. user pushes fixes to broken packages
5. ???
6. merge

Is this what you’re saying (you don’t want people to have to wait longer to find out what to fix), or are you saying that steps 2-4 are repeated since stuff may keep breaking and the user may have to keep pushing?<hr>How about running it as a separate job on new commits to the branches, reusing the first eval output to not block nixpkgs-review or mess with our concurrency too much?

Mic92 replied:
1. user opens PR
2. eval runs
3. user runs nixpkgs-review
4. user pushes fixes to broken packages
5. 2. eval runs
6. user run nixpkgs-review

Relying on github actions is actually a big win over what we had before because we can now also easier test different architectures in parallel with this feature. This would usually than require more than 4 x 16G RAM to reproduce locally if you test all 4 systems we officially support.

Having the second run in a separate action would be already an improvement.

vcunat replied:
Channels: at a quick glance, this check is included in this job https://hydra.nixos.org/job/nixos/unstable-small/nixpkgs.release-checks which is a blocker for all channels/jobsets.

winterqt replied:
> Relying on github actions is actually a big win over what we had before because we can now also easier test different architectures in parallel with this feature. This would usually than require more than 4 x 16G RAM to reproduce locally if you test all 4 systems we officially support.

Right, but you’d still need remote builders for the other three architectures. 😉

> Having the second run in a separate action would be already an improvement.

Alright, I’ll do that for now then.

winterqt replied:
@emilazy and I were discussing this in Matrix, and she noted that doing such a thing is likely bad due to concurrent job limits/our runner allocation limit(s).

So my idea was to have the exact same workflow we do now, but split off the second eval into a workflow that runs after the main eval one finishes, things are compared and tagged, etc. to work around that hurdle.

winterqt replied:
To clarify: by “the four steps,” you mean all the checks? I don’t think it makes sense to run some of them in parallel, but to group them by how likely they are to be resource intensive. e.g. everything except the evals are cheap and quick, so I’d want to keep the current structure but then split off the second eval and outpaths check into a different workflow as described above.

emilazy replied:
> How often did this actually happened in practice that accidentally depend on pkgs.path that justifies that we run every evaluation on every pull request twice?

I’ve seen it within the past few months.

emilazy replied:
See 8353a681b5d6416b4c610dd54f1a176c554ead9b, fixed by 130a10524873af7f4e2bc100cd94ef4be43ce3a5.

The channel getting completely blocked by an eval issue when CI is saying eval is all okay is a terrible contributor experience – I’m sure @ConnorBaker can attest – and we should not let it happen.

ConnorBaker commented on ci/eval/release-checks.nix:
```diff
@@ -0,0 +1,127 @@
+
+let
+  # no point in pulling in another builder for this
+  abortSource = runCommand "dont-use-nix-path-in-nixpkgs.nix" { } ''
+    echo 'builtins.abort "Illegal use of <nixpkgs> in Nixpkgs"' > "$out"
+  '';
```
"We have `echo ... > $out` at home!"

The `echo ... > $out` at home:

```suggestion
  abortSource = builtins.toFile "dont-use-nix-path-in-nixpkgs.nix" ''
    builtins.abort "Illegal use of <nixpkgs> in Nixpkgs"
  '';
```

winterqt replied:
lol, thanks!

ConnorBaker commented on ci/eval/release-checks.nix:
```diff
@@ -0,0 +1,127 @@
+
+    attrpathsPurity = runCommand "attrpaths-purity-check" { } ''
+      if ! diff -u ${attrpaths}/paths.json ${attrpaths'}/paths.json; then
+        echo
+        echo "Error: Nixpkgs evaluation depends on Nixpkgs path"
```
```suggestion
        nixErrorLog "Nixpkgs evaluation depends on Nixpkgs path"
```

uses https://github.com/NixOS/nixpkgs/blob/3c5ac8ecfa2c25e33cbf31c6ddd0f5a9f2acec13/pkgs/stdenv/generic/setup.sh#L109-L113

Imagine I also made this same suggestion a few lines down for the other `echo` followed by `echo "Error:...`

winterqt replied:
I think I'd rather keep the first `echo` so that the diff is separated from the log, wdyt?

ConnorBaker replied:
Since you'll be maintaining it, I trust your judgement about what works best for you -- just wanted to make you aware of `nixErrorLog` if you hadn't seen it since the refactor.

ConnorBaker commented:
Left minor nits/comments which is about all my brain can offer right now; many thanks for working on this!

winterqt committed:
workflows/release-checks: init

winterqt commented:
As of the last commit:
- [Eval](https://github.com/winterqt/nixpkgs/actions/runs/15030866013) took 5m30s
- [Release checks](https://github.com/winterqt/nixpkgs/actions/runs/15030866009) took 6m50s (a good portion of that time spent waiting for eval to finish)

I think we're in a pretty good spot now that both evals run in parallel.

winterqt commented:
I'm not sure why GHA is suddenly running the action from my branch. It wasn't doing so earlier when I switched eval to use release checks, but now it's grabbing the new process job from my branch, causing it to fail? Anyone know why, given it's listed as `pull_request` for `.github/workflows/eval.yml`, and this PR has always touched that; why did it just start now?

winterqt commented:
Ah, I see... it's doing it twice. I guess that makes sense.

winterqt commented:
Right, but I originally thought that it wasn't doing it on my previous pushes, but turns out I just didn't realize it actually was.

winterqt commented:
Oh, I guess I can add that to relchecks to get it to run.

winterqt commented:
Though I guess that begs the question: is GHA doing some weird shenanigans so that #406852 ran in both contexts *before* the PR was merged? I guess it's looking at both the target branch *and* the PR to determine whether to run a job in what contexts?!

winterqt commented on .github/workflows/release-checks.yml:
```diff
@@ -0,0 +1,157 @@
+        with:
+          extra_nix_config: sandbox = true
+
+      - name: Get first eval run id
+        id: targetRunId
+        run: |
+          # Get the latest eval.yml workflow run for the head commit
```
It's the same thing I was doing previously -- it's to make sure `pkgs.path` isn't depended on in eval. `eval.releaseChecks.eval'` forcibly re-evals Nixpkgs from a different store path.

winterqt replied:
Ah, apologies, it seems I forgot to rename the output variable. Look in `env`, `targetRunId` is the run of `github.sha`, so, same action.

winterqt replied:
Ah, seems that `github.sha` *isn't* what I want for the PR case, let me see what would be.

winterqt replied:
`github.sha` is the merge commit, which would be the same as the SHA that triggered `eval.yml` -- why isn't GitHub giving us the job?

winterqt replied:
It was `github.event.head.sha`.

winterqt replied:
Nope, that's pulling in random PRs. What the hell.

winterqt replied:
*sigh*

```diff
-          HEAD_SHA: ${{ github.event.head.sha }}
+          HEAD_SHA: ${{ github.event.pull_request.head.sha }}
```

winterqt commented:
> > Though I guess that begs the question: is GHA doing some weird shenanigans so that #406852 ran in both contexts _before_ the PR was merged? I guess it's looking at both the target branch _and_ the PR to determine whether to run a job in what contexts?!
> 
> Each workflow now triggers on two events. `pull_request_target` as before and `pull_request`. The latter only when the respective workflow file has changed. In this case, the same job is run twice in the PR - once for each context. And yes, that's before merge.

Yes, but now `release-checks` is triggering, so I guess it's also *after* merge?

winterqt commented on .github/workflows/release-checks.yml:
```diff
@@ -0,0 +1,160 @@
+      - .github/workflows/release-checks.yml
+  pull_request_target:
+    types: [opened, synchronize, reopened]
+  push:
+    # Keep this synced with ci/request-reviews/dev-branches.txt
+    branches:
+      - master
+      - staging
+      - release-*
+      - staging-*
+      - haskell-updates
+      - python-updates
```
Why don't we want to do the same on pushes? We're running eval on pushes after all.

winterqt commented:
Trying to figure out why the process job is suddenly complaining, it wasn't on my fork. Rebased fork and testing to see.

winterqt commented:
Failures are expected.

winterqt commented on .github/workflows/release-checks.yml:
```diff
@@ -0,0 +1,150 @@
+      - name: Get first eval run id
+        id: evalRunId
+        run: |
+          # Get the latest eval.yml workflow run for the head commit
+          if ! run=$(gh api --method GET /repos/"$REPOSITORY"/actions/workflows/eval.yml/runs \
+            -f head_sha="$HEAD_SHA" \
+            --jq '.workflow_runs | sort_by(.run_started_at) | .[-1]') \
```
I guess in the case of a PR touching `eval.yml`, this (and the process job) will just grab whichever of the two `eval.yml`s ran first (either `pull_request_target` or `pull_request`), right? It's probably not worth adding any more logic to handle that, though, given how rare it'll be that we make breaking changes to eval like I'm doing here.

winterqt replied:
No, I left it open so I would remember to do it — done.

winterqt replied:
Ah, so this won't work unless the PR that touches relchecks.yml also touches eval.yml... will revert it.

winterqt commented on ci/eval/compare/default.nix:
```diff
@@ -76,10 +76,14 @@ let
     dir:
     let
-      raw = builtins.readFile "${dir}/outpaths.json";
+      # Sorry.
+      raw = runCommand "outpaths.json" { } ''
+        zcat ${dir}/outpaths.json.gz > $out
+      '';
```
I guess I'll comment on this now ~~before I wake up to an angry mob~~: the IFD is required for making the GHA's usage simpler while still keeping good out-of-GHA behavior.

Furthermore, I'd like to eventually migrate Hydra to use the reimplemented release checks (and a soon-to-come reimplemented `tarball` job), but all of those things more or less require the ability for us to pass eval results around without pulling in a comparatively giant closure (or without us losing the benefits of our current setup wrt parallelism between the system evals).

So if we have to make a singular consumer use IFD, I think it's an acceptable tradeoff.

wolfgangwalther reviewed (commented):
> As of the last commit:
>
>    [Eval](https://github.com/winterqt/nixpkgs/actions/runs/15030866013) took 5m30s
>    [Release checks](https://github.com/winterqt/nixpkgs/actions/runs/15030866009) took 6m50s (a good portion of that time spent waiting for eval to finish)
>
> I think we're in a pretty good spot now that both evals run in parallel.

Yes, we are in the right ballpark here, that's great!

We can probably still improve a bit, but we don't need to do that in here, yet. For example, we could probably reduce the waiting, if we don't wait for the whole eval workflow to finish (which includes process and tag), but only up to the point where the intermediate results are available. This should then make release checks not exceed main eval, hopefully.

But as I said, not a blocker, I'll look into that part for main eval anyway at some point.

Mic92 commented:
Long-term this check might be a candidate for merge-queues i.e. only running this check in the merge-queue and not in the pull request. @wolfgangwalther I don't know what your long-term plans are, but I think merge-queues would solve some issues. i.e. we would always have the store paths computed ahead of merging to master and than not run any CI on the master branch and we can have potentially expensive checks run in the merge queue as we can make use of batching multiple pull requests in one go. Also we would no longer have the issue if the CI results are old when we merge things.

Mic92 commented:
> I'm not sure, yet, how merge-queues could avoid running eval on master. We need the comparison against master to set the correct labels and request reviewers, so we can't wait for themerge-queue there.

The merge-queue would compute the comparison before the commits are pushed to master. So you don't need to wait for the comparison to happen because it already happened before the commits are visible in the master branch.

winterqt commented on ci/eval/README.md:
```diff
@@ -8,14 +8,19 @@ nix-build ci -A eval.full \
 - `--cores`: The number of cores to use for each job. Recommended to set this to the amount of cores on your system divided by `--max-jobs`.
 - `chunkSize`: The number of attributes that are evaluated simultaneously on a single core. Lowering this decreases memory usage at the cost of increased evaluation time. If this is too high, there won't be enough chunks to process them in parallel, and will also increase evaluation time.
 - `evalSystems`: The set of systems for which `nixpkgs` should be evaluated. Defaults to the four official platforms (`x86_64-linux`, `aarch64-linux`, `x86_64-darwin` and `aarch64-darwin`).
+- `hydraEmulation`: If `true`, the eval will behave as if it is on Hydra. If `false` (the default), the eval will inherit the same evaluation properties as the release checks run in CI. In order to be 100% sure that a change doesn't break something, you'll need to eval twice for both `hydraEmulation` values (though most of the time, the default mode will suffice).
```
The gist of it is that I’m not really sure if breaking it is really easily possible.

With that being said: once I get master in a better place with regards to attrpath differences per platform for Hydra and non-Hydra, I would like to add some checks to full coverage mode so that we don’t get these mismatches again.

winterqt replied:
To clarify — I would like to keep it, since it is still useful for making sure Hydra and CI are in sync. Not sure what the reactions to that comment imply.

winterqt commented on .github/workflows/eval.yml:
```diff
@@ -53,7 +53,7 @@ jobs:
           MATRIX_SYSTEM: ${{ matrix.system }}
         run: |
-          nix-build nixpkgs/ci -A eval.singleSystem \
+          nix-build nixpkgs/ci -A eval.releaseChecks.eval \
```
Yeah, good point, it doesn’t make much sense to separate them at this point — this was a relic of when I wasn’t ripping out the original eval step.

winterqt commented on .github/workflows/eval.yml:
```diff
@@ -58,11 +58,13 @@ jobs:
             --arg chunkSize 10000
           # If it uses too much memory, slightly decrease chunkSize
 
+          nix-store --export "$(readlink result)" | gzip -n -9 > eval.nar.gz
+
```
Yes, we still pass files around, but this lets the release checks run as easily locally as they do in CI; you don’t need to manually specify eval results as arguments.

It’s an unfortunate consequence of the fact that we have to pass files between machines in the first place, but IMO it’s much more preferable to the alternative of requiring that the eval paths get manually passed to to relchecks.

winterqt replied:
It *does*, but in order to get the behavior between machines that lets the release checks expression define the exact eval expressions it wants to test, we need to import the results into the Nix store. Given we’re outputting store paths in a derivation when evaluating the outpaths, this will naturally pull all of the runCommand dependencies into the result’s closure. While this wasn’t an issue when we just grabbed the files, it is now that we want to export/import from/to the different machines’ stores.

winterqt replied:
We want to keep the definition of the evaluations congruent between the jobs that actually do them and the job that checks them, so we pass around the result of the evaluations as store paths so that Nix can uphold that property for us.

This won’t work with JSON files because as soon as the eval results come out of the builder without obfuscation, Nix will latch onto them and mark those paths (so, `stdenv`, Nix, `jq`, etc.) as dependencies of the results.

When we go to import those onto the other machines, Nix will complain that it doesn’t have access to the dependencies.

To work around this, we obfuscate the outpaths so that the reference scanner doesn’t catch them, which lets us pass the results as store paths between machines without a giant closure coming along with it.

winterqt replied:
Not the host Nix, but the pinned CI Nix, since that’s what we use to eval in drvs. Let me see if I can write a better explanation.

winterqt replied:
So I was able to repro on aarch64-darwin now, which makes things much easier for me. (I have no idea why I wasn't able to before, maybe I was getting some wires crossed.)

The eval drv is defined like this:
```nix
    runCommand "nixpkgs-eval-${evalSystem}"
      {
        nativeBuildInputs = [
          nix
          time
          procps
          jq
        ];
        env = {
          inherit evalSystem chunkSize;
        };
      }
```

Keeping in mind that `nix`, `time`, `procps`, and `jq` all come from *the pinned CI Nixpkgs*, let's look at the result of evaluating Nixpkgs at the current HEAD of this PR's branch:
```
nixVersions.nix_2_24.aarch64-darwin                                         dev=/nix/store/r3b21gjfmcgn46lzng8fk39vwv9brm3m-nix-2.24.14-dev;doc=/nix/store/gp9llp23xil6dx1913178inaghgzyfh9-nix-2.24.14-doc;man=/nix/store/ndgyf3m0pwfad0mzq2h73i4v1fwilqk3-nix-2.24.14-man;/nix/store/88phshibil7l301sjrv84pxgm8cjwzl0-nix-2.24.14
```

And what's the Nix we used in the eval drv?
```
$ nix-store --query --outputs /nix/store/7r8x4r9laj8hm7ymj1pbls73dy9yhk75-nix-2.24.14.drv
/nix/store/88phshibil7l301sjrv84pxgm8cjwzl0-nix-2.24.14
```

So depending on how close the pinned CI Nixpkgs is to the Nixpkgs checkout being eval'd, there's a good chance the eval drvs dependencies will be pulled in as *requisites of the eval output itself*, by virtue of the paths being mentioned, which is why we need to obfuscate them.

winterqt replied:
> Also, the JSON file will **not** contain any references to the pinned tools, no matter how big the drift is.

Yes it will. See above -- right now, the drift is small enough that it does contain the entire stdenv/Nix/jq/etc.

> What exactly is meant by "obfuscating"?

Compressing them obfuscates them so that the reference scanner doesn't catch them.

winterqt replied:
I only somehow just realize the `combine` output isn't something we pass around as a store path, so we can just remove the IFD entirely.

winterqt replied:
> If that was true, why did https://github.com/NixOS/nixpkgs/pull/406909 not show any rebuilds? If the tooling was in the outpaths, and the tooling changes, there must be rebuilds, because it compares against the base branch, before the update of the pin.

Because the thing that rebuilds are being compared against is *the target branch*, which indeed had 0 rebuilds.

Let's use that as an example:
1. I opened that PR against `master`, to update the pinned Nixpkgs.
2. That PR changed the Nixpkgs we use in CI to be only a few days behind `master`, so the Nix & co from the target branch were now what we were using to eval.
3. However, it did not touch Nixpkgs itself -- there were still **zero rebuilds** compared to the target branch, the *only thing* that changed was the Nix we use to eval said Nixpkgs. But since the Nix in the Nixpkgs we're evaling is now the same as the Nix we're using to eval, the eval output's closure *contains the Nix that we're using to eval*.

Does that make sense? 😅

winterqt replied:
> Another nix with hash BBBB, because it's part of the tooling (as above, we disagree). If the duplicate was removed, that's another variation with the same outcome.

No, *there is no duplicate*. The paths are leaking into the JSON only because the pinned tools are the same as the ones in the Nixpkgs being eval'd.

winterqt replied:
Let me clarify:

> nix with hash AAAA, because it's part of the tooling that creates the file (that's what you're saying, I disagree)

You're correct, this would *not* happen.

> Another nix with hash BBBB, because it's part of the tooling (as above, we disagree). If the duplicate was removed, that's another variation with the same outcome.

No, there is only one Nix with hash `BBBB`. However, since Nix with hash `BBBB` is used within the eval drv, Nix scans `$out/paths`, notices it has `BBBB` in it, and treats it as a dependency of the file.

winterqt replied:
So you're right that normally it wouldn't matter, but for the release checks we're explicitly using the intermediate results and not the JSON file.

But as you saw me realize, we can just remove the IFD and only make this an issue for what matters -- the intermediate results. Let me clean that up after I'm done with this rustc bump.

winterqt commented on ci/eval/default.nix:
```diff
@@ -65,7 +70,8 @@ let
       # because `--argstr system` would only be passed to the ci/default.nix file!
       evalSystem,
       # The path to the `paths.json` file from `attrpathsSuperset`
-      attrpathFile ? "${attrpathsSuperset}/paths.json",
+      attrpathFile ? "${attrpathsSuperset { evalSystem = attrpathEvalSystem; }}/paths.json",
+      attrpathEvalSystem ? "x86_64-linux",
```
The default was switched in that commit just to split up and provide a motivation for the change that allows you to set it -- of course, it's moot with the following changes, but I think the separation is a nice thing to show how the idea(s) involved. I can always squash if you'd rather.

winterqt commented:
Okay, *now* we're good. Hopefully I didn't mess any of the commits up, please let me know if it looks like I did.

winterqt commented:
I’ve gone over these commits at least twice from the last push and it looks like I didn’t screw anything up in terms of the splits/atomicness etc.

So we should be good to go, unless there are any concerns about commit descriptions/documentation/etc.

(And I’d still hopefully like to hear from @ConnorBaker about the log thing before merge.)

winterqt commented on .github/workflows/release-checks.yml:
```diff
@@ -0,0 +1,149 @@
+          github-token: ${{ github.token }}
+          run-id: ${{ steps.evalRunId.outputs.evalRunId }}
+
+      - name: Import results
+        if: steps.evalRunId.outputs.evalRunId
+        run: |
+          for path in first-store-paths/*/* second-store-paths/*/*; do
+            nix-store --import < "$path"
+          done
```
Right - I wouldn’t mind making this simpler, though I have one concern about the API surface: how can we keep `releaseChecks` compartmentalized so that we can use the `eval'` without needing to pass some dummy version of it?

I suspect if I just keep how it’s done now, Nix’s laziness will do the work for me, but let me play around with it.

winterqt replied:
I answered my own question, dw about it.

winterqt commented on ci/eval/default.nix:
```diff
@@ -65,7 +69,7 @@ let
       evalSystem,
       # The path to the `paths.json` file from `attrpathsSuperset`
-      attrpathFile ? "${attrpathsSuperset}/paths.json",
+      attrpathFile ? "${attrpathsSuperset { inherit evalSystem; }}/paths.json",
```
I’m pretty sure this would cause no issues for nixpkgs-review users, but I’m going to take a closer look at it to be sure.

It’s definitely just a byproduct of the attrpath switch (as you also noticed), but I’m also not a fan of the idea of splitting this into separate PR that blocks this one. So let me verify what exactly is going on (I already have an idea) and come back to it.

winterqt replied:
Alright, this is now fixed. Let's talk about the remaining "rebuilds."

Previously, we included all packages marked as `broken`. Now, we don't. This is why you see over 3,000 removals.

For the additions (well, most of them, there's too many on `x86_64-linux` for me to go through without going insane):
- All of the Android SDK stuff is things that we were just straight up not picking up on before due to evaluating the attrpaths on `aarch64-linux`. (We were before the ARM runner migration. 😉)
- `box86` is special in that its `meta.platforms` does not at all represent what platforms it can actually run on, but thanks to these changes, we pick up on it now. [Even Hydra doesn't do that!](https://hydra.nixos.org/job/nixpkgs/trunk/box86.aarch64-linux)
- `coreboot-utils`'s `meta.platforms` isn't correctly applied, so it doesn't actually throw on Darwin (when it should).
- `elf-header`, `libintl`, and `tests.pkg-config.defaultPkgConfigPackages.hidapi` are cursed packages that eval to `null` on the default Linux platforms.
- `zoom-us` wasn't included in the attrpaths post-ARM runners, as it's littered with a bunch of `throw`s.
- `sidequest` doesn't have its `meta` set correctly.

Most of these are probably fixable, but I'm not going to do them in this PR.

---

That begs the question: what's going to happen to PRs that aren't rebased onto this branch, will their comparisons be screwed up? They'd probably just see a bunch of "removals," so hopefully not. After merge, I probably will do something like ping all committers in a pinned issue so they're aware.

winterqt replied:
Ah, right, I forgot we always use the target branch... so this should be fine.

An aside: wow, GitHub doesn't even bother to remove the commit that was picked from the PR view, lol.

winterqt committed:
pkgs/top-level/release-outpaths: explicitly enable unfree packages

This codifies what we've been doing by setting `includeBroken` to `true`,
which is evaluating unfree packages anyways. I don't see a reason to not
eval them, so this makes it so that we can do it when we set `includeBroken`
to false in the next commit.

winterqt commented:
@wolfgangwalther We probably want to backport to both 24.11 and 25.05 since we'll still be supporting/running CI on 24.11 for a month.

winterqt commented:
> I removed the automatically applied backport label for 24.11 here, because this label would still be present on the 25.05 branch. Thus CI changes would be marked for backport to 25.05 on master and then for backport to 24.11 on the backport PR.

Ah.

winterqt commented on .github/workflows/eval.yml:
```diff
@@ -91,6 +93,12 @@ jobs:
+            evalPath=$(nix-store --import < "$path")
+            cp -r "$evalPath" intermediate # combine doesn't care about the directory names
+          done
+
```
It's gone.

winterqt commented on .github/workflows/release-checks.yml:
```diff
@@ -0,0 +1,142 @@
+  pull_request:
+    paths:
+      - .github/workflows/release-checks.yml
+  pull_request_target:
```
Might be a good idea to keep this in sync with `eval.yml` -- that is, trigger on undrafts -- just so we can pick up and test any new merge commit that `eval.yml` is evaluating.

winterqt replied:
Why didn’t we just do that in the first place, out of curiosity? I was under the impression that it just wasn’t possible since we didn’t do it, given it’s a pretty obvious optimization. ;)

That may be a Silvan question and not you, though.

winterqt replied:
Also, wait, doesn’t that have the downside of not testing a later merge, leading CI to be green when it shouldn’t be? Maybe we *should* keep the current behavior for now, in lieu of a proper merge queue?

winterqt commented on ci/eval/default.nix:
```diff
@@ -186,7 +210,22 @@ let
 
         # Make sure the glob doesn't break when there's no files
         shopt -s nullglob
```
With the addition of `jq` into the pipeline, is this even needed anymore? Like, I’m pretty sure jq will error if it’s given no input.

winterqt replied:
Ah, indeed, didn't notice the stats removal.

winterqt commented on pkgs/top-level/release.nix:
```diff
@@ -62,6 +62,16 @@
+  #
+  # Like `attrNamesOnly`, the behavior of this flag may
+  # change at any time.
+  includeUnsupported ? false,
```
I need to update this comment to reflect the slightly tweaked behavior.

winterqt replied:
> Instead it's an addition on top, correct?

No, that’s not right, unfortunately.

How release checks eval Nixpkgs is different from how we evaluate it (using `pkgs/top-level/release-outpaths.nix`), so in order to get parity, we need to make this change, so that we *attempt* to evaluate all packages on all platforms, even if they’re marked as unsupported. Without this change, those packages won’t even get included in the attrpaths set at all.

In general I think this split in evaluation techniques is… iffy at best, and I’d like to eventually go over it with a fine-toothed comb and see if there’s improvements that can be made, but now is not that time.

winterqt replied:
> Ah.. ok. It's because of the two step process of "collect attrpaths" - "evaluate attrpaths", right?

Right.

> So on the first step, even with no allowUnfree in sight anywhere, those would already be evaluated up until they throw "unfree!". But then they are removed from the attrpaths... and the actual eval job never gets there.

Not quite -- this isn't related to unfree packages; unfree packages are eval'd before this PR because of both `includeBroken = true` and `allowUnfree = false` being set. `includeUnsupported` is so that we try to evaluate packages on every platform, even unsupported ones, just to make sure that it gets to the "unsupported!" throw.

(The `includeBroken = false` changed is why originally there were a ton of unfree packages "removed" in this PR, it's because it stopped the (I guess unintentional, given what you and Silvan have said...) evaluation of unfree packages. So that's why we then have the change that enables it, to get as much of the same behavior as possible.)

wolfgangwalther reviewed (commented):
From what I can tell this is in good shape now, only minor suggestions for improvement. That's with the exception of the `allowUnfree` change where I am unsure, because of lack of knowledge about the hydra setup.

winterqt commented on pkgs/top-level/release-outpaths.nix:
```diff
@@ -27,7 +27,7 @@ let
           config = {
             allowAliases = false;
             allowBroken = includeBroken;
-            allowUnfree = false;
+            allowUnfree = true;
```
`release-outpaths.nix` is only used by CI, not Hydra.

winterqt commented:
Trying to figure out what went wrong with the eval purity check... doesn't seem to be an actual issue (e.g. we're not actually dependent on `pkgs.path`).

winterqt commented:
Ah! So the issue is that the `get-merge-commit` jobs got different merge commits, which obviously will fail. I guess we need to store the merge commit used for the first one, but I guess there's a chance that won't even exist by the time the second job runs?

winterqt commented:
Aaaaand they're different again... I cannot win.

winterqt commented:
Let me hack on this, I have an idea.

winterqt committed:
workflows/get-merge-commit: only grab merge commit once

Currently, the `get-merge-commit` job is inherently racey, as we can't
control whether these jobs will run at the same time, so it's possible
that the merge/target commits will change if the target gets updated
in between `get-merge-commit` runs.

This is bad enough in theory (and maybe has even caused some oddities in
the CI setup in the past), but it essentially makes the release checks
untentable on days where lots of people are merging things.

So, this change fixes it by electing the eval workflow as the "leader,"
and having all other jobs wait for it to determine which commits will be
checked.

winterqt commented:
> ... does this mean we effectively need to put both into the same workflow, because otherwise they are not guaranteed to have the same merge commit?

Nah, I'm trying out [`4b56ca2` (#406825)](https://github.com/NixOS/nixpkgs/pull/406825/commits/4b56ca2480cf908a8987c2cf0ea84c8424a64706) as a fix.

> Maybe it's effectively the same merge commit, but with a different hash, because it's recreated for every workflow with a slightly different timestamp?

It's not, it's recreated every time you reload the merge status by refreshing the PR page, or Whenever GitHub Feels Like It.

winterqt commented on .github/workflows/get-merge-commit.yml:
```diff
@@ -32,6 +40,7 @@ jobs:
           sparse-checkout: ci
 
       - name: Check if the PR can be merged and get the test merge commit
+        if: inputs.leader == true
```
I know these conditions look stupid, but I genuinely could not get it to work otherwise. lol.

winterqt commented on .github/workflows/get-merge-commit.yml:
```diff
@@ -44,14 +53,62 @@ jobs:
+
+      - name: Get leader workflow's run ID
+        id: getLeaderRun
+        timeout-minutes: 5
```
I think this is fine for a failure mode -- it's unlikely that we need to add anything else like a check to see if the eval workflow failed as a stopgap...? *shrug*

winterqt commented:
As for why the relchecks job is failing now: we're playing with fire and RNG because of `pull_request` vs `pull_request_target`, but as we've already discussed, it's likely not worth fixing.

winterqt commented on ci/eval/default.nix:
```diff
@@ -69,11 +75,22 @@ let
       # because `--argstr system` would only be passed to the ci/default.nix file!
       evalSystem,
       # The path to the `paths.json` file from `attrpathsSuperset`
-      attrpathFile ? "${attrpathsSuperset { inherit evalSystem; }}/paths.json",
+      attrpathFile ? "${attrpathsSuperset { inherit evalSystem nixpkgsPath; }}/paths.json",
+      nixpkgsPath ? nixpkgs,
       # The number of attributes per chunk, see ./README.md for more info.
       chunkSize,
       checkMeta ? true,
-      includeBroken ? true,
+
+      # Don't try to eval packages marked as broken.
+      includeBroken ? false,
```
Split.

winterqt replied:
Please let me know if these commit messages make sense -- I think I've stared at these changes for long enough that I may be writing nonsensically/incorrectly. :-)

winterqt commented on .github/workflows/get-merge-commit.yml:
```diff
@@ -5,6 +5,14 @@ on:
     paths:
       - .github/workflows/get-merge-commit.yml
```
Might want to consider removing this now that this requires inputs (and apparently defaults don’t apply to non-`workflow_run` triggers…).

winterqt commented:
> The `get-merge-commit` workflow has increased massively in complexity. I feel like this took a wrong turn after the latest discovery, that different workflows for a single event are not guaranteed to run on the same merge commit (which still puzzles me).

Right, but I think this *is* a desirable change for consistency, mainly avoiding weird results in the kind of conditions we just observed. I’d be willing to split it into a separate PR. Let me know.

Regarding your idea: I like it in theory, but I’ll need to think about it a bit more.

emilazy commented:
Does that work if the problem is that the existence of attribute paths in some way depends on the value of `pkgs.path`? (If the actual release checks don’t catch this then I suppose it’s fine, though. But IIRC they do?)

philiptaron commented:
At one point I was tracking with the back-and-forth on this PR, but I've lost the plot. Where are we with this effort, @winterqt and @wolfgangwalther?

philiptaron commented:
Good news: I'm the maintainer of `nixpkgs-vet`. I'd be delighted to see it grow checks for those sorts of defects. I agree it's the right place to add them.

I'll have time to work on that in a few weeks.
